ensime_shared/protocol.py

```diff
@@ -232,10 +232,10 @@ def handle_doc_uri(self, call_id, payload):
 
     def handle_completion_info_list(self, call_id, payload):
         """Handler for a completion response."""
+        self.log.debug('handle_completion_info_list: in')
         completions = payload["completions"]
-        self.log("handle_completion_info_list: in")
         self.suggestions = [completion_to_suggest(c) for c in completions]
-        self.log("handle_completion_info_list: {}".format(self.suggestions))
+        self.log.debug('handle_completion_info_list: %s', self.suggestions)
```

Comment:
I think you're referring generally to data structure output in cases where it's a large structure.

The advantage with the old-style `%s` format strings is that the `logging` module elides the actual string formatting if the log level isn't active, so CPU cycles aren't needlessly spent. The `str` representations are fine in many cases like the sample output above IMO, but for larger messages like this where it may be warranted the optimal thing to do is something like this:

```python
if self.log.isEnabledFor(logging.DEBUG):
    self.log.debug('handle_completion_info_list:\n%s', pprint.pformat(self.suggestions))
```

Slightly more intrusive code, but worth it if the output is otherwise pretty useless.

I'm happy to update cases like this in the PR if you all help point them out, I haven't taken the time to consider/check the output of each case too closely, beyond a quick assessment of what level is appropriate.